### DISCUSSION TITLE
is it possible to use the embedded fonts in a pdf file to write another pdf?

### DISCUSSION THREAD

**Discussion by leyiwang** [pymupdf/PyMuPDF#1829] [Q&A] [answered]
I have a word document opened via office, then i save this docx file as a pdf file. i open this pdf file with pymupdf and extract the embedded fonts, then i write the text info to a new pdf file with these fonts. the code is as follows:
```Python
import fitz
doc = fitz.open("a.pdf")
pdf = fitz.open()
for page in doc:
    fonts = page.get_fonts()
    pos = 90
    for font in fonts:
        new_page = pdf.new_page()
        font_name, _type, _, buffer = doc.extract_font(font[0])
        new_page.insert_font(fontname=font_name, fontbuffer=buffer)
        print(font_name)
        new_page.insert_text(fitz.Point(pos, pos), text="hello, 你好", fontname=font_name)
pdf.save("new.pdf")
```
when i execute this code , contents in each page of the new pdf seems to be wrong. i am very confused, is it possible to use the embedded fonts in a pdf file to write another pdf?

[a.pdf](https://drive.google.com/file/d/1G_GAdPzSmrdQeW98EkFUxh_0_Xmnow7Q/view?usp=sharing)
[new.pdf](https://drive.google.com/file/d/1FTGLfm21ql2SXRRPoaWGMbsGwPIOQOK3/view?usp=sharing)

**Comment by JorjMcKie** [comment]
This is no issue, but a typical post for the "Discussions" tab.

**Comment by JorjMcKie** [comment, marked as answer]
>  is it possible to use the embedded fonts in a pdf file to write another pdf?

Not as easily as it may seem:
The fonts created by Word's export to PDF are only font **_subsets_**: they contain **_exactly_** the characters that are used in the docx - no more. So you cannot write the word "hello" with them because this word's characters are not contained in the font subset.

To put it in simple words: you probably can only write the same same text again with each character in its original font.
But even then: the way Word stores its font subsets prevents doing this either, because Word also changes the mapping of unicode to glyph (= visual appearance) number, when creating the font subset. 
So if you want to write "实" again using the extracted subset font, the glyph number it is giving you for "实" may be wrong (and probably **will be** wrong) and garbage (or nothing) comes out.

**Reply by leyiwang** [reply]
i see, thank you for reply.
